keps/sig-storage/20191031-storage-capacity-constraints-for-pod-scheduling.md

```diff
@@ -0,0 +1,894 @@
+storage, there will be one `CSIStoragePool` per CSI
+[Topology](https://github.com/container-storage-interface/spec/blob/4731db0e0bc53238b93850f43ab05d9355df0fd9/lib/go/csi/csi.pb.go#L1662-L1691)
+and driver.
+
+### Size of ephemeral inline volumes
+
+Currently Kubernetes has no information about the size of an ephemeral
+inline volume. A new `CSIVolumeSource.fsSize` field needs to be added
+to expose that in a vendor-agnostic way. Details for that are in
+https://github.com/kubernetes/enhancements/pull/1353.
+
+### Pod scheduling
+
+The Kubernetes scheduler monitors `CSIStoragePool` objects and
+excludes nodes with insufficient remaining capacity when it comes to
+making scheduling decisions for a pod which uses ephemeral inline
+volumes or persistent volumes with delayed binding. If the driver does
+not indicate that it supports capacity reporting, then the scheduler
+proceeds just as it does now, so nothing changes for existing CSI
+driver deployments.
+
+### API
+
+All Kubernetes API changes are under the `CSIStorageCapacity` feature
+gate.
+
+#### CSIStoragePool
+
+The prefix “CSI” is used to avoid confusion with other types that have
+a similar name and because some aspects are specific to CSI.
+
+```
+// CSIStoragePool identifies one particular storage pool and
+// stores the corresponding attributes. The spec is read-only.
+//
+// There should never be two objects with the same spec. If
+// a consumer of this information finds two objects with the
+// same spec, it should use the information from the one with the
+// most recent CreationTimestamp.
+type CSIStoragePool struct {
+	metav1.TypeMeta
+	// Standard object's metadata. The name has no particular meaning and just has to
+	// meet the usual requirements (length, characters, unique). To ensure that
+    // there are no conflicts with other CSI drivers on the cluster, the recommendation
+    // is to use sp-<uuid>.
+	//
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+
+	Spec   CSIStoragePoolSpec
+	Status CSIStoragePoolStatus
+}
+
+// CSIStoragePoolSpec contains the constant attributes of a CSIStoragePool.
+//
+// A pool might only be accessible from a subset of the nodes in the
+// cluster. That subset can be identified either via NodeTopology or
+// NodeList, but not both. If neither is set, the pool is assumed
+// to be available in the entire cluster.
+type CSIStoragePoolSpec struct {
+    // The CSI driver that provides access to the storage pool.
+	// This must be the string returned by the CSI GetPluginName() call.
+	DriverName string
+
+	// NodeTopology can be used to describe a storage pool that is available
+	// only for nodes matching certain criteria.
+	// +optional
+	NodeTopology *v1.NodeSelector
+
+	// Nodes can be used to describe a storage pool that is available
+	// only for certain nodes in the cluster.
+	//
+	// +listType=set
+	// +optional
+	Nodes []string
```

Comment:
According to your local storage example [here](https://github.com/kubernetes/enhancements/pull/1353/files/#diff-f3298a4cee270f1b16246ce3b5504ab1R383), you're using this as a simple version of `NodeTopology`? I don't think this is necessary, because 1) they are basically identical (`Nodes` can be translated into a selector with `kubernetes.io/hostname` as the key and `nodes` as the value), 2) [local PV](https://kubernetes.io/docs/concepts/storage/volumes/#local) uses`NodeSelector` too, I think it's ok to use NodeSelector for local storage.

Reply:
`Nodes` is indeed redundant. But it is considerably shorter than a full-blown `NodeSelector`, which makes `Nodes` easier to read (see examples) and also more efficient (shorter JSON). Therefore I think it's worthwhile to have it.

I also had considered to have just a `Node string` field. But the advantage of `Node: somehost` over `Nodes: [somehost]` is a lot smaller, so I decided against that.

Reply:
if Nodes is an array here, then, I don't think its possible to inform capacity per node, as the CSIStoragePoolStatus is having capacity per StorageClass.
For ex., if Nodes here is node1, node2, i.e.,
```
nodes:
- node1
- node2
```
and status is:
```
status:
  classes:
  - capacity: 256G
    storageClassName: striped
  - capacity: 128G
    storageClassName: mirrored
```
we have no way of passing capacity available per node.
So, here, its better to be 'Node' rather than 'Nodes'

Reply:
In this case we need two `CSIStoragePool` objects, one with `nodes: [ node 1 ]`  and another one with `nodes: [ node2 ]`.

Reply:
thanks for clarifying

Reply:
Done, see 1fdd83d250e1817075d95e4db08840a704ee3a88.